src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+        (some block-height)
+        none
+        none
+        zonefile-hash)
+      (ok true))))
+
+;; NAME_UPDATE
+;; A NAME_UPDATE transaction changes the name's zone file hash. You would send one of these transactions 
+;; if you wanted to change the name's zone file contents. 
+;; For example, you would do this if you want to deploy your own Gaia hub and want other people to read from it.
+(define-public (name-update (namespace (buff 19))
+                            (name (buff 16))
+                            (zonefile-hash (buff 20)))
+  (let (
+    (owner (unwrap!
+      (nft-get-owner? names (tuple (name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+    (name-props (unwrap!
+      (map-get? name-properties ((name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND)))) ;; The name must exist
+    ;; The sender must match the name's current owner
+    (asserts!
+      (is-eq owner contract-caller)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The name must not be in the renewal grace period
+    (asserts!
+      (is-eq (unwrap! (is-name-in-grace-period? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_GRACE_PERIOD))
+    ;; The name must not be expired 
+    (asserts!
+      (is-eq (unwrap! (is-name-lease-expired? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_EXPIRED))
+    ;; The name must not be revoked
+    (asserts!
+      (is-none (get revoked-at name-props))
+      (err ERR_NAME_REVOKED))
+    ;; Update the zonefile
+    (update-zonefile-and-props
+      namespace 
+      name  
+      (get registered-at name-props)
+      (get imported-at name-props)
+      none
+      zonefile-hash)
+    (ok true)))
+
+;; NAME_TRANSFER
+;; A NAME_TRANSFER transaction changes the name's public key hash. You would send one of these transactions if you wanted to:
+;; - Change your private key
+;; - Send the name to someone else
+;; When transferring a name, you have the option to also clear the name's zone file hash (i.e. set it to null). 
+;; This is useful for when you send the name to someone else, so the recipient's name does not resolve to your zone file.
+(define-public (name-transfer (namespace (buff 19))
+                              (name (buff 16))
+                              (new-owner principal)
+                              (zonefile-hash (optional (buff 20))))
+  (let (
+    (current-owned-name (map-get? owner-name ((owner new-owner))))
+    (namespace-props (unwrap!
+      (map-get? namespaces ((namespace namespace)))
+      (err ERR_NAMESPACE_NOT_FOUND))))
+    (let (
+      (owner (unwrap!
+        (nft-get-owner? names (tuple (name name) (namespace namespace)))
+        (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+      (name-props (unwrap!
+        (map-get? name-properties ((name name) (namespace namespace)))
+        (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+      (can-new-owner-get-name (if (is-none current-owned-name)
```

Comment:
The `can-new-owner-get-name` method might be better implemented as a `define-read-only` function, so wallets can determine if a recipient is even able to receive the name before sending a transaction.

Reply:
Also, this check is not complete -- if the user revoked a name, then it no longer owns it, and should be able to receive a new name.  A principal can own at most one non-revoked name (but it can have many revoked names).

Reply:
Good idea. Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/19a6ffc687c7118230d4249df93cae24cea3d536